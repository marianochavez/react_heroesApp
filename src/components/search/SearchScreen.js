import { useNavigate, useLocation } from "react-router-dom";
import queryString from "query-string";
import { getHeroesByName } from "../../helpers/getHeroesByName";
import { useForm } from "../../hooks/useForm";
import { HeroCard } from "../hero/HeroCard";
import { useMemo } from "react";

export const SearchScreen = () => {

  const navigate = useNavigate();
  const location = useLocation();

  const { q = '' } = queryString.parse(location.search);

  const [formValues, handleInputChange] = useForm({
    searchText: q,
  });

  const { searchText } = formValues;
  
  const heroesFiltered = useMemo( () => getHeroesByName(q), [q]);

  const handleSearch = (e) => {
    e.preventDefault();
  navigate(`?q=${searchText}`);
  };

  return (
    <>
      <h1>Searchs</h1>
      <hr/>

      <div className="row">
        <div className="col-5">
          <h4>Search Form</h4>
          <hr/>
          <form onSubmit={handleSearch}>
            <input
              type='text'
              placeholder="Search heroe"
              className="form-control"
              name="searchText"
              autoComplete="off"
              value={searchText}
              onChange={handleInputChange}
            />
            <button className="btn btn-outline-primary mt-3" type="submit">
              Buscar...
            </button>
          </form>
        </div>
        <div className="col-7">
          <h4>Resultados</h4>
          <hr/>
          {
            (q==='')
              ? <div className="aler alert-info animate__animated animate__flash">
                Busca un heroe!
              </div>
              : ( heroesFiltered.length === 0 ) 
                && <div className="alert alert-danger animate__animated animate__flash">
                  No se encontro: {q}
                </div>
          }
          {
            heroesFiltered.map(hero => (
              <HeroCard key={hero.id} hero={hero}/>
            ))
          }
        </div>
      </div>
    </>
  );
};
