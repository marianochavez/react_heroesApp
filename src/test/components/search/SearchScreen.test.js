import {mount} from "enzyme";
import { MemoryRouter } from "react-router-dom";
import { SearchScreen } from "../../../components/search/SearchScreen";

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate,
}));

describe('SearchScreen tests', () => {
  
    test('should display correctly', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/']}>
                <SearchScreen />
            </MemoryRouter>
        );  
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Busca un heroe!');
    });

    test('should display Batman and input queryString', () => {
      
        const wrapper = mount(
            <MemoryRouter initialEntries={['/?q=Batman']}>
                <SearchScreen />
            </MemoryRouter>
        );
        expect(wrapper.find('input').prop('value')).toBe('Batman');
        expect(wrapper.find('.card-title').text().trim()).toBe('Batman');
        expect(wrapper).toMatchSnapshot();
    });

    test('should display error if query doesnt exists', () => {
      
        const wrapper = mount(
            <MemoryRouter initialEntries={['/?q=Batman123']}>
                <SearchScreen />
            </MemoryRouter>
        );
        expect(wrapper.find('input').prop('value')).toBe('Batman123');
        expect(wrapper.find('.alert').text().trim()).toBe('No se encontro: Batman123');
    });
    
    test('should call navigate on the new screen', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', {
            target: {
                name: 'searchText',
                value: 'Batman'
            }
        });

        wrapper.find('form').prop('onSubmit')({
            preventDefault: () => {}
        });

        expect(mockNavigate).toHaveBeenCalledWith('?q=Batman');
    });

    
    
    
});
