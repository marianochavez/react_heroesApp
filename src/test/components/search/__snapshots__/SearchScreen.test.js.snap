// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchScreen tests should display Batman and input queryString 1`] = `
Array [
  <h1>
    Searchs
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search heroe"
          type="text"
          value="Batman"
        />
        <button
          className="btn btn-outline-primary mt-3"
          type="submit"
        >
          Buscar...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="col-4 animate__animated animate__fadeIn"
      >
        <div
          className="card mb-3"
        >
          <div
            className="row g-0"
          >
            <div
              className="col-4"
            >
              <img
                alt="Batman"
                className="card-img"
                src="/assets/dc-batman.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  className="cart-text"
                >
                  <small
                    className="text-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  See more
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`SearchScreen tests should display correctly 1`] = `
Array [
  <h1>
    Searchs
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search heroe"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-3"
          type="submit"
        >
          Buscar...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="aler alert-info animate__animated animate__flash"
      >
        Busca un heroe!
      </div>
    </div>
  </div>,
]
`;
